### ISSUE TITLE
CSL modifies transaction when deserializing from CBOR

### ISSUE THREAD

**Issue by TotallyNotChase** [Emurgo/cardano-serialization-lib#550]
It seems like CSL is modifying a transaction during deserialization, thereby changing its transaction id. Furthermore, re-serializing it (without any modifications) yields a different cbor. I wanted to ask whether this behavior was intentional, and if so - what was the motivation for it.

Consider the tx cbor:

```
84a6008382582016deb843715068524ce94bd405172207422e9318d03021c73b2896c17adc22440182582074294f2903f50367ff3c0d94d965ee607aa149cda4511980e98506339ab7c34d01825820a4880b9decc7c09439b1d3feab0267ecd3dc1865d57e6bddca9562e089892c4c000d81825820898095a5eacc401df98fa5dd18d87b42acbc7935d6b8bf429b1de3603c34281b010182a200583900e2f9d92651c75a28717bf5622e6164e25133d856e9c02ea21a234dfc91a0730692d58ba66b472db59deaf6d4d4ea2bebc3c8257c2431defc011b000000025252b0a4a200583900e2f9d92651c75a28717bf5622e6164e25133d856e9c02ea21a234dfc91a0730692d58ba66b472db59deaf6d4d4ea2bebc3c8257c2431defc01821a0029d3a8ab581c005eaf690cba88f441494e42f5edce9bd7f595c56f99687e2fa0aad4a1447443485301581c201e808091c0c75ba59d3074d751e7be2e7a0ce68da21e870a46ccfca1447443485301581c2d636c02b9bed0f81597e9ff33a46106924835ebee862ed4d47a07cfa1447443485301581c71b3165770ebb46654224e10860b257e32f19ba2ba217f9b48e266f9a1447443485301581c7a3c470398fd61369d684161d3687539ea0e42124f4163f5a3273a8da1447443485301581c88b0001389e7986daedbceafea7813949d67c5401047febb167f8ca7a1447443485301581ca142a46e64519978ccd98b1736d6e01cbeede252a1b10403f5690cdaa1447443485301581ca74ee162ab940f63e389e81cee5f4335b41fa4c59e128be0cc3837f0a1447443485301581cc5a2c612f8647019b765ebadd8bee79144cd28a685f0309a67690da4a14474474d5201581cea17af2bbe9582cdda7f216ca6b8cfd48a65b093aa23aed8fe11d971a1447443485301581cee122c1a8ba777396466790e93e07d50646f794a01676e31b102d96ca1447443485301021a0008b7ae09a1581ca142a46e64519978ccd98b1736d6e01cbeede252a1b10403f5690cdaa14474434853010b582078ca17a1749fbb77dc1753491278466ad2540a14232d1a40aafd078d1af89586a203815908d05908cd01000033332332232332232323232323232323232323232323232323232222223232323235500222222222225335333553024120013232123300122333500522002002001002350012200112330012253350021001102d02c25335325335333573466e3cd400488008d404c880080b40b04ccd5cd19b873500122001350132200102d02c102c3500122002102b102c00a1326320223357389201115554784f206e6f7420636f6e73756d65640002315335333573466e3c048d5402c880080ac0a854cd4ccd5cd19b8701335500b2200102b02a10231326320223357389210c77726f6e6720616d6f756e740002313263202233573892010b77726f6e6720746f6b656e00023135500122222222225335330245027007162213500222253350041335502d00200122161353333573466e1cd55cea8012400046644246600200600464646464646464646464646666ae68cdc39aab9d500a480008cccccccccc888888888848cccccccccc00402c02802402001c01801401000c008cd40548c8c8cccd5cd19b8735573aa0049000119910919800801801180f1aba15002301a357426ae8940088c98c809ccd5ce01381401289aab9e5001137540026ae854028cd4054058d5d0a804999aa80c3ae501735742a010666aa030eb9405cd5d0a80399a80a80f1aba15006335015335502101f75a6ae854014c8c8c8cccd5cd19b8735573aa00490001199109198008018011919191999ab9a3370e6aae754009200023322123300100300233502475a6ae854008c094d5d09aba2500223263202b33573805605805226aae7940044dd50009aba150023232323333573466e1cd55cea8012400046644246600200600466a048eb4d5d0a80118129aba135744a004464c6405666ae700ac0b00a44d55cf280089baa001357426ae8940088c98c809ccd5ce01381401289aab9e5001137540026ae854010cd4055d71aba15003335015335502175c40026ae854008c06cd5d09aba2500223263202333573804604804226ae8940044d5d1280089aba25001135744a00226ae8940044d5d1280089aba25001135744a00226aae7940044dd50009aba150023232323333573466e1d400520062321222230040053016357426aae79400c8cccd5cd19b875002480108c848888c008014c060d5d09aab9e500423333573466e1d400d20022321222230010053014357426aae7940148cccd5cd19b875004480008c848888c00c014dd71aba135573ca00c464c6403c66ae7007807c07006c0680644d55cea80089baa001357426ae8940088c98c805ccd5ce00b80c00a9100109aab9e5001137540022464460046eb0004c8004d5406488cccd55cf8009280c119a80b98021aba100230033574400402446464646666ae68cdc39aab9d5003480008ccc88848ccc00401000c008c8c8c8cccd5cd19b8735573aa004900011991091980080180118099aba1500233500c012357426ae8940088c98c8058cd5ce00b00b80a09aab9e5001137540026ae85400cccd5401dd728031aba1500233500875c6ae84d5d1280111931900919ab9c012013010135744a00226aae7940044dd5000899aa800bae75a224464460046eac004c8004d5405c88c8cccd55cf8011280b919a80b19aa80c18031aab9d5002300535573ca00460086ae8800c0444d5d080089119191999ab9a3370ea0029000119091180100198029aba135573ca00646666ae68cdc3a801240044244002464c6402066ae700400440380344d55cea80089baa001232323333573466e1cd55cea80124000466442466002006004600a6ae854008dd69aba135744a004464c6401a66ae7003403802c4d55cf280089baa0012323333573466e1cd55cea800a400046eb8d5d09aab9e500223263200b33573801601801226ea8004488c8c8cccd5cd19b87500148010848880048cccd5cd19b875002480088c84888c00c010c018d5d09aab9e500423333573466e1d400d20002122200223263200e33573801c01e01801601426aae7540044dd50009191999ab9a3370ea0029001100911999ab9a3370ea0049000100911931900519ab9c00a00b008007135573a6ea80048c8c8c8c8c8cccd5cd19b8750014803084888888800c8cccd5cd19b875002480288488888880108cccd5cd19b875003480208cc8848888888cc004024020dd71aba15005375a6ae84d5d1280291999ab9a3370ea00890031199109111111198010048041bae35742a00e6eb8d5d09aba2500723333573466e1d40152004233221222222233006009008300c35742a0126eb8d5d09aba2500923333573466e1d40192002232122222223007008300d357426aae79402c8cccd5cd19b875007480008c848888888c014020c038d5d09aab9e500c23263201233573802402602001e01c01a01801601426aae7540104d55cf280189aab9e5002135573ca00226ea80048c8c8c8c8cccd5cd19b875001480088ccc888488ccc00401401000cdd69aba15004375a6ae85400cdd69aba135744a00646666ae68cdc3a80124000464244600400660106ae84d55cf280311931900599ab9c00b00c009008135573aa00626ae8940044d55cf280089baa001232323333573466e1d400520022321223001003375c6ae84d55cf280191999ab9a3370ea004900011909118010019bae357426aae7940108c98c8020cd5ce00400480300289aab9d5001137540022244464646666ae68cdc39aab9d5002480008cd5403cc018d5d0a80118029aba135744a004464c6401066ae700200240184d55cf280089baa001499240103505431001200132001355008221122253350011350032200122133350052200230040023335530071200100500400132001355007222533500110022213500222330073330080020060010033200135500622225335001100222135002225335333573466e1c005200000d00c1333008007006003133300800733500912333001008003002006003112200212212233001004003112212330010030021220021220011123230010012233003300200200148009221047443485300335122330024892016deb843715068524ce94bd405172207422e9318d03021c73b2896c17adc224400480088848cc00400c00880050581840100d87980821a00326cf41a3c19c72df5f6
```

It is somewhat of a simple babbage era minting transaction, though it contains a plutus script. The transaction was built and serialized using the official cardano tooling: [`cardano-api`](https://github.com/input-output-hk/cardano-node/tree/1.35.3/cardano-api) at 1.35.3.

Deserializing this CBOR and then obtaining its id:

```js
import * as C from '@emurgo/cardano-serialization-lib-browser'

function hexToByteArray(hex) {
    let bytes = [];
    for (let c = 0; c < hex.length; c += 2) {
        bytes.push(parseInt(hex.substr(c, 2), 16));
    }
    return new Uint8Array(bytes);
};

function byteArrayToHex (bytes) {
    return Array.from(bytes, x => x.toString(16).padStart(2, '0')).join('');
}

let txCborHex = '...';

const tx = C.Transaction.from_bytes(hexToByteArray(txCborHex));
const txHash = C.hash_transaction(tx.body());

> txHash.to_hex()
'5c10c18731e8cd88042a37a11d3cb81ba4a998d19f82c251dfd43510efe2fb70'
```

However, if I use cardano-api to deserialize the same and obtain its id:

```hs
import qualified Cardano.Api as Api
import Data.ByteString (ByteString)
import qualified Data.ByteString.Base16 as B16

> txCborHex = "..." :: ByteString

> bs = either error id $ B16.decode txCborHex
> Api.getTxId . Api.getTxBody <$> Api.deserialiseFromCBOR (Api.AsTx Api.AsBabbageEra) bs
Right "1635c684923db4a4a30d3f5ae96c2dd22638d77cf5a9baaa9dd71029bf05c4e8"
```

Indeed, they are different.

In fact, the transaction itself seems to have been inexplicably changed during deserialization by CSL.

```js
byteArrayToHex(tx.to_bytes())
```

gives:

```
84a6008382582016deb843715068524ce94bd405172207422e9318d03021c73b2896c17adc22440182582074294f2903f50367ff3c0d94d965ee607aa149cda4511980e98506339ab7c34d01825820a4880b9decc7c09439b1d3feab0267ecd3dc1865d57e6bddca9562e089892c4c00018282583900e2f9d92651c75a28717bf5622e6164e25133d856e9c02ea21a234dfc91a0730692d58ba66b472db59deaf6d4d4ea2bebc3c8257c2431defc1b000000025252b0a482583900e2f9d92651c75a28717bf5622e6164e25133d856e9c02ea21a234dfc91a0730692d58ba66b472db59deaf6d4d4ea2bebc3c8257c2431defc821a0029d3a8ab581c005eaf690cba88f441494e42f5edce9bd7f595c56f99687e2fa0aad4a1447443485301581c201e808091c0c75ba59d3074d751e7be2e7a0ce68da21e870a46ccfca1447443485301581c2d636c02b9bed0f81597e9ff33a46106924835ebee862ed4d47a07cfa1447443485301581c71b3165770ebb46654224e10860b257e32f19ba2ba217f9b48e266f9a1447443485301581c7a3c470398fd61369d684161d3687539ea0e42124f4163f5a3273a8da1447443485301581c88b0001389e7986daedbceafea7813949d67c5401047febb167f8ca7a1447443485301581ca142a46e64519978ccd98b1736d6e01cbeede252a1b10403f5690cdaa1447443485301581ca74ee162ab940f63e389e81cee5f4335b41fa4c59e128be0cc3837f0a1447443485301581cc5a2c612f8647019b765ebadd8bee79144cd28a685f0309a67690da4a14474474d5201581cea17af2bbe9582cdda7f216ca6b8cfd48a65b093aa23aed8fe11d971a1447443485301581cee122c1a8ba777396466790e93e07d50646f794a01676e31b102d96ca1447443485301021a0008b7ae09a1581ca142a46e64519978ccd98b1736d6e01cbeede252a1b10403f5690cdaa14474434853010b582078ca17a1749fbb77dc1753491278466ad2540a14232d1a40aafd078d1af895860d81825820898095a5eacc401df98fa5dd18d87b42acbc7935d6b8bf429b1de3603c34281b01a203815908d05908cd01000033332332232332232323232323232323232323232323232323232222223232323235500222222222225335333553024120013232123300122333500522002002001002350012200112330012253350021001102d02c25335325335333573466e3cd400488008d404c880080b40b04ccd5cd19b873500122001350132200102d02c102c3500122002102b102c00a1326320223357389201115554784f206e6f7420636f6e73756d65640002315335333573466e3c048d5402c880080ac0a854cd4ccd5cd19b8701335500b2200102b02a10231326320223357389210c77726f6e6720616d6f756e740002313263202233573892010b77726f6e6720746f6b656e00023135500122222222225335330245027007162213500222253350041335502d00200122161353333573466e1cd55cea8012400046644246600200600464646464646464646464646666ae68cdc39aab9d500a480008cccccccccc888888888848cccccccccc00402c02802402001c01801401000c008cd40548c8c8cccd5cd19b8735573aa0049000119910919800801801180f1aba15002301a357426ae8940088c98c809ccd5ce01381401289aab9e5001137540026ae854028cd4054058d5d0a804999aa80c3ae501735742a010666aa030eb9405cd5d0a80399a80a80f1aba15006335015335502101f75a6ae854014c8c8c8cccd5cd19b8735573aa00490001199109198008018011919191999ab9a3370e6aae754009200023322123300100300233502475a6ae854008c094d5d09aba2500223263202b33573805605805226aae7940044dd50009aba150023232323333573466e1cd55cea8012400046644246600200600466a048eb4d5d0a80118129aba135744a004464c6405666ae700ac0b00a44d55cf280089baa001357426ae8940088c98c809ccd5ce01381401289aab9e5001137540026ae854010cd4055d71aba15003335015335502175c40026ae854008c06cd5d09aba2500223263202333573804604804226ae8940044d5d1280089aba25001135744a00226ae8940044d5d1280089aba25001135744a00226aae7940044dd50009aba150023232323333573466e1d400520062321222230040053016357426aae79400c8cccd5cd19b875002480108c848888c008014c060d5d09aab9e500423333573466e1d400d20022321222230010053014357426aae7940148cccd5cd19b875004480008c848888c00c014dd71aba135573ca00c464c6403c66ae7007807c07006c0680644d55cea80089baa001357426ae8940088c98c805ccd5ce00b80c00a9100109aab9e5001137540022464460046eb0004c8004d5406488cccd55cf8009280c119a80b98021aba100230033574400402446464646666ae68cdc39aab9d5003480008ccc88848ccc00401000c008c8c8c8cccd5cd19b8735573aa004900011991091980080180118099aba1500233500c012357426ae8940088c98c8058cd5ce00b00b80a09aab9e5001137540026ae85400cccd5401dd728031aba1500233500875c6ae84d5d1280111931900919ab9c012013010135744a00226aae7940044dd5000899aa800bae75a224464460046eac004c8004d5405c88c8cccd55cf8011280b919a80b19aa80c18031aab9d5002300535573ca00460086ae8800c0444d5d080089119191999ab9a3370ea0029000119091180100198029aba135573ca00646666ae68cdc3a801240044244002464c6402066ae700400440380344d55cea80089baa001232323333573466e1cd55cea80124000466442466002006004600a6ae854008dd69aba135744a004464c6401a66ae7003403802c4d55cf280089baa0012323333573466e1cd55cea800a400046eb8d5d09aab9e500223263200b33573801601801226ea8004488c8c8cccd5cd19b87500148010848880048cccd5cd19b875002480088c84888c00c010c018d5d09aab9e500423333573466e1d400d20002122200223263200e33573801c01e01801601426aae7540044dd50009191999ab9a3370ea0029001100911999ab9a3370ea0049000100911931900519ab9c00a00b008007135573a6ea80048c8c8c8c8c8cccd5cd19b8750014803084888888800c8cccd5cd19b875002480288488888880108cccd5cd19b875003480208cc8848888888cc004024020dd71aba15005375a6ae84d5d1280291999ab9a3370ea00890031199109111111198010048041bae35742a00e6eb8d5d09aba2500723333573466e1d40152004233221222222233006009008300c35742a0126eb8d5d09aba2500923333573466e1d40192002232122222223007008300d357426aae79402c8cccd5cd19b875007480008c848888888c014020c038d5d09aab9e500c23263201233573802402602001e01c01a01801601426aae7540104d55cf280189aab9e5002135573ca00226ea80048c8c8c8c8cccd5cd19b875001480088ccc888488ccc00401401000cdd69aba15004375a6ae85400cdd69aba135744a00646666ae68cdc3a80124000464244600400660106ae84d55cf280311931900599ab9c00b00c009008135573aa00626ae8940044d55cf280089baa001232323333573466e1d400520022321223001003375c6ae84d55cf280191999ab9a3370ea004900011909118010019bae357426aae7940108c98c8020cd5ce00400480300289aab9d5001137540022244464646666ae68cdc39aab9d5002480008cd5403cc018d5d0a80118029aba135744a004464c6401066ae700200240184d55cf280089baa001499240103505431001200132001355008221122253350011350032200122133350052200230040023335530071200100500400132001355007222533500110022213500222330073330080020060010033200135500622225335001100222135002225335333573466e1c005200000d00c1333008007006003133300800733500912333001008003002006003112200212212233001004003112212330010030021220021220011123230010012233003300200200148009221047443485300335122330024892016deb843715068524ce94bd405172207422e9318d03021c73b2896c17adc224400480088848cc00400c00880050581840100d87980821a00326cf41a3c19c72df5f6
```

Which is different from the original CBOR.

Deserializing and then reserializing using cardano-api gives me the same exact CBOR however, which seems logical.

**Comment by lisicky** [comment]
Hi @TotallyNotChase! It is possible that CBOR after deserialization and serialization would be different because we don't store the original format for all structs in the CSL. For example, a map or an array or a tx output can be serialized differently with the same data. Array or map can be with definite and indefinite length, tx output can be serialized in Babbage or pre-Babbage format. And if you deserialize only for read-only purposes, it's better to use the original cbor if you already signed the tx or used the tx hash somewhere before.

**Comment by TotallyNotChase** [comment]
I suspected it was due to the ambiguity within arrays/maps/similar. However, this still does seem quite problematic to me.

The usecase currently sort of requires interfacing with CSL really. Unlike in the example, CSL _is_ actually used to modify the witness set and add signers to a transaction (with finalized tx body) returned by cardano-api. The issue is that the transaction id is changing due to this discrepancy. So the transaction id returned by cardano-api, for example, would no longer be correct once CSL is used to, say - add witnesses to the transaction.

Not the end of the world to me, but this sort of ambiguity really does make cross language communication harder, sometimes implicitly so.
